### ISSUE TITLE
bug with filter/grouping

### ISSUE THREAD

**Issue by dtnguyen2** [sherpa/sherpa#149]
Mike McCollough reported a bug while running specfit.  I was able to trace the problem to the following the code of get_staterror in the file sherpa/astro/data.py.  In this particular case, specfit is normally with a notice region and a requirement of a minimum of 16 counts per bin.  As a result, within the aforementioned function: filter is True so apply_filter (which internally calls apply_grouping) returns an array with one of the elements < 16 (the last element of the cnts array: 0) which is obviously a bug.  I suspect that once this bug is fixed, then a lot of our tests will fail.  I pity the poor soul that has to fix this bug.

```
        if filter:
            cnts = self.apply_filter(cnts)
        else:
            cnts = self.apply_grouping(cnts)
```

the cnts array looks like the following, note the last 0

(Pdb) print cnts
[ 18.  16.  17.  24.  16.  19.  23.  22.  17.  17.  19.  16.  17.  16.  19.
1.  16.  19.  16.  16.  16.  17.  16.  16.  16.  16.  16.   0.]

In short there is something wrong with the filter/group

**Comment by jbudynk** [comment]
@dtnguyen2 and I were investigating the issue, and it doesn't look like a bug anymore.

If using `group_counts(num)`, it is possible for the last array in a grouped dataset to have less than `num`. 

If we reach the end of a dataset, and the summed counts for the last bin are less than `num`, the group is marked as incomplete and given a quality of 2, meaning they're "bad" and should be ignored in analysis (See [extern/grplib/grplib.c::grp_do_num_counts()](https://github.com/sherpa/sherpa/blob/master/extern/grplib-4.7/src/grplib.c#L130) at around line 139). 

One can ignore these poor quality points with `sherpa.astro.ui.ignore_bad()` before analysis.

For example:

```
from sherpa.astro.ui import *

# load the 3c273.pi file in sherpa-test-data/sherpatest/
load_pha('3c273.pi')
group(16)
print get_data().to_fit()[0]

[17.0, 16.0, 17.0, 16.0, 18.0, 21.0, 17.0, 23.0, 18.0, 21.0,
  22.0, 21.0, 19.0, 21.0, 17.0, 17.0, 17.0, 17.0, 21.0, 17.0,
  20.0, 17.0, 18.0, 17.0, 18.0, 17.0, 16.0, 16.0, 17.0, 17.0,
  17.0, 16.0, 16.0, 17.0, 17.0, 16.0, 17.0, 16.0, 17.0, 16.0,
  16.0, 9.0]

# now ignore the bad data
ignore_bad()
print get_data().to_fit()[0]

[17.0, 16.0, 17.0, 16.0, 18.0, 21.0, 17.0, 23.0, 18.0, 21.0,
  22.0, 21.0, 19.0, 21.0, 17.0, 17.0, 17.0, 17.0, 21.0, 17.0,
  20.0, 17.0, 18.0, 17.0, 18.0, 17.0, 16.0, 16.0, 17.0, 17.0,
  17.0, 16.0, 16.0, 17.0, 17.0, 16.0, 17.0, 16.0, 17.0, 16.0,
  16.0]
```

I've added some simple group_counts() tests to the branch "bug-#149-filter-grouping" so that we at least have a test of what group_counts() should be doing. If this is *not* the expected behavior of group_counts(), let me know and I'll update the tests as necessary.

**Comment by jbudynk** [comment]
However, there was an issue raised when using get_counts() or get_dep(). Say I have a simple dataset:

```
# make a dataset of 161 "1"s
x = numpy.arange(start=1, stop=161, step=1)
y = numpy.ones(x.size)
load_arrays(0, x, y, ui.DataPHA)

# grouping these to 16 should mean I have an array of length 10 with each element equal to 16.
group_counts(0,16)

# control arrays
newx = numpy.arange(start=1, stop=161, step=16)
newy = numpy.zeros(newx.size)+16

# I expect the grouped data to look like this
print newy
[ 16.  16.  16.  16.  16.  16.  16.  16.  16.  16.]

# but instead I get
print get_counts(0)
[ 1.  1.  1.  1.  1.  1.  1.  1.  1.  1.]
```

Granted, this could easily be myself not understanding something, but it doesn't make much sense to me... I've traced the code down to [sherpa.astro.data.py::_fix_y_units()](https://github.com/sherpa/sherpa/blob/bug-%23149-filter-grouping/sherpa/astro/data.py#L1415). Commenting out `val /= numpy.abs(ebin)` (line 1456) gives the expected results.

**Comment by dtnguyen2** [comment]
As I have reported before, if specfit is run with with the default ngrp=16
then sherpa returns the following data array (note the last entry is 0).

newdevel12-113: specfit.py
infile=/data/L3/mccolml/rel2/new_spectra/test_runs/data/test_100/acisf01878_000N001_r0100_pha3.fits.gz
outfile=fubar.par error=none model=powerlaw
[ 18.  16.  17.  24.  16.  19.  23.  22.  17.  17.  19.  16.  17.  16.  19.
  17.  16.  19.  16.  16.  16.  17.  16.  16.  16.  16.  16.   0.]

As suggested by Jamie, I added the ignore_bad command so rectify the
aforementioned problem.  Unfortunately, this is the output now.  I have
already shown this to Jamie a few weeks ago, but I just want to send it
here for the record before we forget that there is still an issue with
sherpa-group

newdevel12-117: specfit.py
infile=/data/L3/mccolml/rel2/new_spectra/test_runs/data/test_100/acisf01878_000N001_r0100_pha3.fits.gz
outfile=fubar.par error=none
model=powerlaw

[  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.  18.  16.  17.  24.  16.  19.  23.  22.  17.  17.
19.

  16.  17.  16.  19.  17.  16.  19.  16.  16.  16.  17.  16.  16.  16.
16.

  16.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   1.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   1.   0.
0.

   0.   0.   1.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
0.

   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.]

On Wed, Feb 3, 2016 at 5:03 PM, Jamie Budynkiewicz <notifications@github.com

**Comment by jbudynk** [comment]
An interesting note: using ui.ignore_bad() in the specfit code, the output is correct after the 31st element (0-base), and up to the last good quality bin. The first 32 elements are filled with the ungrouped counts values. I tested to see if this is true with a few other files:

```
python specfit.py infile=/export/scialg/sherpa/sherpa-test-data/sherpatest/acisf01575_001N001_r0085_pha3.fits.gz outfile=fubar.par
[  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   1.   1.   2.   2.   0.
   2.   0.   4.   4.  19.  18.  16.  16.  17.  26.  19.  18.  20.  20.  17.
  17.  19.  17.  22.  17.  19.  16.  16.  19.  23.  19.  16.  16.  19.  18.
  20.  17.  20.  16.  17.  19.  16.  18.  16.  18.  16.  16.  17.  16.  17.
  16.  16.  16.  16.  16.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   1.   1.   0.   1.
...
   0.   0.   0.   0.   0.   0.   0.   0.   1.]

# file in @dtnguyen2's comment above
python specfit.py infile=/data/L3/mccolml/rel2/new_spectra/test_runs/data/test_100/acisf01878_000N001_r0100_pha3.fits.gz outfile=fubar.par
[  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.  18.  16.  17.  24.  16.  19.  23.  22.  17.  17.  19.
  16.  17.  16.  19.  17.  16.  19.  16.  16.  16.  17.  16.  16.  16.  16.
  16.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
...
   0.   0.   0.   0.   0.]

/acisf02257_000N001_r0007_pha3.fits.gz
python specfit.py infile=/data/L3/mccolml/rel2/new_spectra/test_runs/data/test_100/acisf02257_000N001_r0007_pha3.fits.gz outfile=fubar.par
[  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   1.
   1.   0.   0.   0.   1.   0.   0.   0.   1.   1.   1.   2.   0.   4.   1.
   2.   2.   2.   4.  18.  16.  19.  16.  26.  19.  17.  26.  25.  18.  17.
  22.  20.  19.  16.  22.  19.  17.  17.  16.  18.  20.  16.  16.  16.  16.
  16.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.
   0.   0.   0.   0.   0.   1.   0.   0.   0.   0.   0.   0.   0.   0.   0.
...
   0.   0.   0.   0.   0.]
```

This last file produces very interesting results, in that the counts before the 31st element add up to 16 before the 31st element. The result should be

```
[ 17.  16.  24.  19.  16.  26.  19.  17.  26.  25.  18.  17.  22.  20.  19.
  16.  22.  19.  17.  17.  16.  18.  20.  16.  16.  16.  16.  16.  17.  16.
  16.]
```

Note: to get the output to print out from specfit, it's in `group_filter_data()`. Insert print statements  code near line 484, after the `apply_grouping(id, ngrp, tabStops=invmask)` call.

**Comment by jbudynk** [comment]
Hmm... nevermind about this statement:

> the output is correct after the 31st element (0-base), and up to the last good quality bin

This clearly isn't true for `acisf02257_000N001_r0007_pha3.fits.gz`, the last test in the comment above. There are several valid bins at the end of the data array that weren't counted.

**Comment by jbudynk** [comment]
`ignore_bad` removes previous filters. That's why `ignore_bad()` doesn't work in specfit; the data is filtered before it is grouped, so `ignore_bad()` gives the unfiltered results seen above. However, without ignoring the bad quality grouped bins, the last element in the array will always be there ([see the doc notes](https://github.com/sherpa/sherpa/blob/master/sherpa/astro/data.py#L808-L815)).

**Comment by anetasie** [comment]
Can you try doing `ignore_bad()`  before `group()`?  Would this work as expected?

**Comment by olaurino** [comment]
Pending confirmation from @DougBurke and @anetasie I would like to close this bug.

The main issue here seemed to be that the result of a grouping operation, in particular when grouping by counts, should not include any incomplete groups (e.g. at the end of the file). The problem with this statement is that there seems to be conflicting expectations about what filtering and grouping commands should do. The `dmgroup` library clearly states that if a group is not complete its channels are set a quality flag of 2<sup>1</sup>. That seems to imply that one should be able to use `ignore_bad()` to get rid of them. However `ignore_bad()` claims it will remove filters if data is grouped and filtered, which is not always what happens in practice, as shown in #416, as in some cases `ignore_bad()` fulfills the semantics of simply ignoring the incomplete groups.

Also, team scientists requested that the incomplete groups be left in the data set to be fit by `specfit`, which seems to invalidate the idea that returning the last, incomplete group is a bug. Users should get that group and then decide what to do with it. It might be a bug that the latest group can sometimes be 0, which is something we dealt with directly in `specfit`, but I am not even sure about that, because the 0 group covers channels in the range defined by the user, so there may be information in that 0. Removing the 0 group would reduce the range over which the data array is defined, breaking the original filter set by the user. Again, lacking specific requirements, it is probably up to the user, as we do in specfit, to deal with the final 0, the inclusion or exclusion of which may change the fit results.

So, I created two new bugs, #416 and #417. I claim they cover the issues we found during the development of specfit and we hadn't already captured in a ticket (e.g. #233 and #234). If @anetasie and @DougBurke agree we can close this issue.

Footnotes
--------------
<sup>1</sup> From http://cxc.harvard.edu/ciao/ahelp/dmgroup.html:
>For each grouping option, a group may be left incomplete due to insufficient counts or an insufficient number of rows. This may occur because of a tab interval, the end of the dataset, or the presence of a previously grouped row. All rows in such an incomplete group will be given a quality flag=2. For example, for the option NUM_BINS with the grouptypeval=4 and an input file with 26 rows, the resulting 4 groups would be (in the case of no tab intervals): group 1 (rows 1-6), group 2 (rows 7-12), group 3 (rows 13-18), and group 4 (rows 19-24), with rows 25 and 26 ungrouped and given quality flags=2.

**Comment by DougBurke** [comment]
Note that the documented `ignore_bad` behavior is likely added to try to match the observed behavior, rather than being a specification of the behavior, but that can now be discussed in another issue!

I am happy to close this issue, but I haven't been following the details that closely.

**Comment by olaurino** [comment]
Closing. If @anetasie disagrees we can reopen it.

**Comment by anetasie** [comment]
I noticed that this issue has  been closed, but I'm wondering where do we keep track of `ignore_bad` issues.

**Comment by olaurino** [comment]
@anetasie you can find the relevant information in the comment above https://github.com/sherpa/sherpa/issues/149#issuecomment-351769243 as well as in the emails I sent to our internal mailing list and our meeting notes.
